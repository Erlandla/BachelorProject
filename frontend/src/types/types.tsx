export enum Status {
  newProblem = 'Nytt problem',
  started = 'Påbegynnt',
  solved = 'Løst',
}

export interface User {
  email: string
  telephone: string
  affiliation: string
}

export interface ContextUser {
  email: string
  password: string
  affiliation: string
  telephone: string
  isLoggedIn: boolean
  isAdmin: boolean
}

export interface ProblemCardProps {
  id: string
  title: string
  vendor: string
  status: Status
  specificProblem: string
  clearDataProduct: string
  accessibleData: string
  definedAction: string
  subCount: number
  owner: User
  subs: User[]
  edit: boolean
  approved: boolean
}

export interface IfetchType {
  limit: number
  searchPhrase: string
  categoryFilter: string
  orderBy: string
  email?: string
  relation?: number
  approved?: boolean
  similarProblem?: string
  filter?: number
}

export interface IfetchTypeMyProblems {
  limit: number
  searchPhrase: string
  categoryFilter: string
  orderBy: string
  email?: string
  relation?: number
}

export interface Categories {
  specificProblem: string[]
  accessibleData: string[]
  dataProduct: string[]
}

export interface AdminInfo {
  title: string
  count: number
  link: string
}
