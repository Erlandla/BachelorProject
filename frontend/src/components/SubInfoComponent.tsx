import React from 'react'
import { type User } from '../types/types'
import LocationCityIcon from '@mui/icons-material/LocationCity'
import EmailIcon from '@mui/icons-material/Email'
import PhoneIcon from '@mui/icons-material/Phone'
import validateUser from '../utils/validateUser'

function SubInfoComponent(props: User) {
  try {
    validateUser(props)
  } catch (error: any) {
    const errorMsg = 'En feil oppstod: '.concat(error)
    return (
      <div>
        <p className="color-red items-center">{errorMsg}</p>
      </div>
    )
  }

  return (
    <div className="flex flex-col flex-wrap gap-1 text-white justify-between bg-buttonDark py-2.5 px-2.5 text-xs sm:flex-row sm:text-base">
      <div className="flex flex-row items-center gap-1">
        <LocationCityIcon sx={{ fontSize: '1rem' }} />
        {props.affiliation}
      </div>
      <div className="flex flex-row items-center gap-1">
        <EmailIcon sx={{ fontSize: '1rem' }} />
        {props.email}
      </div>
      <div className="flex flex-row items-center gap-1">
        <PhoneIcon sx={{ fontSize: '1rem' }} />
        +47 {props.telephone}
      </div>
    </div>
  )
}

export default SubInfoComponent
