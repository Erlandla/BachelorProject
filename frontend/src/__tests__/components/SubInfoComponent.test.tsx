import React from "react";
import { cleanup, render, screen } from '@testing-library/react';
import { User } from "../../types/types";
import SubInfoComponent from "../../components/SubInfoComponent";
import renderer from "react-test-renderer";

const userValidInfo: User = {
    email: "testmail@testkommune.no",
    telephone: "111 22 333",
    affiliation: "Test kommune"
};
const userInvalidMail: User = {
    email: "hmmmmmmmm",
    telephone: "111 22 333",
    affiliation: "Test kommune"
};
const userInvalidPhone: User = {
    email: "testmail@testkommune.no",
    telephone: "11",
    affiliation: "Test kommune"
};

beforeEach(() => {
    cleanup();
});

it("Renders with valid props data", () => {
    render(
        <SubInfoComponent 
            email={userValidInfo.email} 
            telephone={userValidInfo.telephone} 
            affiliation={userValidInfo.affiliation} 
        />
    );
    expect(screen.getByText("+47 111 22 333")).toBeInTheDocument();
});

it("Throws an error on invalid email", () => {
    render(
        <SubInfoComponent 
            email={userInvalidMail.email} 
            telephone={userInvalidMail.telephone} 
            affiliation={userInvalidMail.affiliation} 
        />
    )
    expect(screen.getByText("En feil oppstod: Error: Ugyldig email"))
});

it("Throws an error on invalid phone number", () => {
    render(
        <SubInfoComponent 
            email={userInvalidPhone.email} 
            telephone={userInvalidPhone.telephone} 
            affiliation={userInvalidPhone.affiliation} 
        />
    );
    expect(screen.getByText("En feil oppstod: Error: Ugyldig telefonnummer"))
});

it("Matches snapshot", () => {
    const snapshot = renderer.create(
        <SubInfoComponent 
            email={userValidInfo.email}
            telephone={userValidInfo.telephone}
            affiliation={userValidInfo.affiliation}
        />
    );
    expect(snapshot).toMatchSnapshot();
})