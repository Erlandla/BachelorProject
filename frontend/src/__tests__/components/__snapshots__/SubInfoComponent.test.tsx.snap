// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="flex flex-col flex-wrap gap-1 text-white justify-between bg-buttonDark py-2.5 px-2.5 text-xs sm:flex-row sm:text-base"
>
  <div
    className="flex flex-row items-center gap-1"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
      data-testid="LocationCityIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M15 11V5l-3-3-3 3v2H3v14h18V11h-6zm-8 8H5v-2h2v2zm0-4H5v-2h2v2zm0-4H5V9h2v2zm6 8h-2v-2h2v2zm0-4h-2v-2h2v2zm0-4h-2V9h2v2zm0-4h-2V5h2v2zm6 12h-2v-2h2v2zm0-4h-2v-2h2v2z"
      />
    </svg>
    Test kommune
  </div>
  <div
    className="flex flex-row items-center gap-1"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
      data-testid="EmailIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
      />
    </svg>
    testmail@testkommune.no
  </div>
  <div
    className="flex flex-row items-center gap-1"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
      data-testid="PhoneIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M6.62 10.79c1.44 2.83 3.76 5.14 6.59 6.59l2.2-2.2c.27-.27.67-.36 1.02-.24 1.12.37 2.33.57 3.57.57.55 0 1 .45 1 1V20c0 .55-.45 1-1 1-9.39 0-17-7.61-17-17 0-.55.45-1 1-1h3.5c.55 0 1 .45 1 1 0 1.25.2 2.45.57 3.57.11.35.03.74-.25 1.02l-2.2 2.2z"
      />
    </svg>
    +47 
    111 22 333
  </div>
</div>
`;
